### ISSUE TITLE
Incremental delivery of gargantuan amounts of audio data from OfflineAudioContext rendering

### ISSUE THREAD

**Issue by JohnWeisz** [WebAudio/web-audio-api#2445]
**Describe the issue**

Currently, `OfflineAudioContext.startRendering` will unconditionally and _always_ result in the creation of an AudioBuffer containing the rendered audio, regardless of whether it is actually used or not.

This is a problem when used to render a massive duration, as it results in a gargantuan memory impact. Take the following settings for example:

 - Sample rate: 48kHz
 - Channels: 4
 - Duration: 4 hours

Rendering this monster, we have `48000 * 4 * 4 * 60 * 60 = 2_764_800_000` float32 values total, which amounts to more than **11 *Gigabytes*** of in-memory data (more specifically, we don't, because no end-user browser will allow this allocation without flags, some even _with_ flags, not speaking of systems which don't even have this much RAM in the first place).

**Where Is It**

https://webaudio.github.io/web-audio-api/#dom-offlineaudiocontext-startrendering, more specifically in step 4 of the OfflineAudioContext.startRendering process:

 > Create a new AudioBuffer, with a number of channels, length and sample rate equal respectively to the numberOfChannels, length and sampleRate values passed to this instance’s constructor in the contextOptions parameter. Assign this buffer to an internal slot [[rendered buffer]] in the OfflineAudioContext.

**Additional Information**

To avoid the above issue, OfflineAudioContext should "somehow" allow skipping this preliminary _forced_ memory allocation, and be able to deliver data at regular intervals.

The way I see it, there are several options here, and I personally think the [MediaRecorder.ondataavailable](https://w3c.github.io/mediacapture-record/#dom-mediarecorder-ondataavailable) is a good starting point, which is an already established example of incremental data delivery.

 1. Having OfflineAudioContext expose a ReadableStream that's pushed as data is available from rendering 
    - _Note: this has the potential to synergize _extremely_ well with the proposed [WebCodecs API](https://github.com/WICG/web-codecs/blob/master/explainer.md), which has a proposal to work out of a ReadableStream -- if the OfflineAudioContext exposed a ReadableStream, it would be possible to plug it into the WebCodecs API directly and acquire formatted audio files in an efficient and elegant manner, which is a very strong point for this option_
 2. Emitting some sort of a `dataavailable` event, possibly with an AudioBuffer for each event containing a chunk of the rendered data
    - _Probably unnecessary, as AudioWorklet is a viable alternative already_
 3. Rendering to a Blob (I personally don't see how the details would work out here)
 4. Simply not returning _anything_, and instead letting user code work it out using AudioWorklets (emitting custom "events", building a stream, etc.)

**Polyfilling**

While the memory impact is unavoidable with the current implementations of startRendering, for some options above, a polyfill based on an AudioWorkletNode (or even ScriptProcessorNode) could easily make the incremental data delivery an option early on, allowing for prototyping the API in advance and allowing progressive upgrades as the change lands in browsers.

**Comment by boourns** [comment]
I have a related, but slightly different use case for breaking up OfflineAudioContext rendering.  If it would be better in a separate issue please let me know and I'll move it.

I have a web application that simultaneously generates audio and reactive video.  The video stream is often analyzing the audio stream to generate interesting visuals.  When running in realtime, an AudioContext runs to generate the audio and the video stream is generated frame by frame in a  `requestAnimationFrame` loop.

It would be nice to be able to run the final audio and video render offline to generate higher quality visuals without any loss of frames or loss of synchronization, even on slower hardware.  My naive request would be a way to run the following loop for the duration of the piece:

- render 1/60th of a second of audio in the `OfflineAudioContext`
- render a frame of video (which may depend on the `OfflineAudioContext` state, for example the results of an `AnalyzerNode` perhaps)

My naive API request would be to add a `duration` parameter to `OfflineAudioContext.startRendering()` specified in frames or in time.  Alternately if the streaming proposal above had a way to pause the OfflineAudioContext until the video frame is completed, that would potentially work as well as we could render a video frame after N audio frames have been delivered.

**Comment by AshleyScirra** [comment]
I think incremental delivery of offline rendered audio is something we've also run in to with [Construct Animate](https://www.construct.net/en/blogs/construct-official-blog-1/announcing-new-product-beta-1589).

We want to offline encode videos with audio for animation purposes. First of all there's a strange mismatch between Web Audio and WebCodecs: with WebCodecs we can do video encoding by drawing to a canvas frame-by-frame and get encoded video data as we go along. We also suspend and resume the OfflineAudioContext frame-by-frame as we go along to make sure things like procedurally-started audio playback gets included. However we still only get the full rendered audio track right at the end. This also forces us to buffer all the encoded video frames until the very end so they can be interleaved for muxing, which worsens the memory use problem (unless we add in buffering video frames to storage, which adds complexity and reduces performance).

I think perhaps an overlooked solution to this is to return rendered audio buffers with the existing `suspend()` method. It currently resolves to void. It could resolve to the audio buffer rendered so far.

For example:

```js
// Schedule a suspension at 10 seconds
audioContext.suspend(10)
.then(audioBuffer => HandlePartialRender(audioBuffer));

// Start rendering up to the 10 second mark
audioContext.startRendering();

// This can then be repeated by calling suspend() with a new time,
// then resume()
```

Then by repeatedly suspending and resuming the OfflineAudioContext, you could get chunks of the rendered audio in whatever time increments you want.

The only other thing to solve then is to make sure OfflineAudioContext doesn't pre-allocate a full-length buffer upon creation. One way to do that might be to make the length parameter optional: if not provided it doesn't preallocate any buffer. Then it would only create as big a buffer as necessary to reach the next scheduled suspend time, and resolve with that buffer when `suspend()` resolves. (Starting rendering with no suspend scheduled would throw in this mode.) Then the caller can get chunks of offline rendered audio as it goes along, on any time period it wants - frame-by-frame, in 1-minute chunks, or whatever, only using as much memory as that time period requires.

This approach has a limited API surface and should work naturally with the existing suspend/resume model of offline rendering.

**Comment by guest271314** [comment]
To stream with Web Audio API use `MediaStreamAudioDestinationNode` or `AudioWorklet`. Using a series of `AudioBuffer` objects is suboptimal.

Here we stream a 200MB+ file https://github.com/guest271314/AudioWorkletStream.

With `MediaStreamAudioDestinationNode` we can just connect to the underlying `MediaStreamTrack`.

SInce we can stream to `MediaStreamAudioDestinationNode`, we can write `AudioData` to a `MediaStreamTrackGenerator`, incrementally, and stream indefinitely. Chrome has a bug where `MediaStreamAudioDestinationNode` does not output silence by default, so we use an `OsciallatorNode` outputting silence to start reading the stream therefrom with `MediaStreamTrackProcessor` and utilize the `timestamp` output from silent `AudioData` objects that we use later when constructing `AudioData` from raw PCM after converting interweved to planar, then write to the `Writable` `MediaStreamTrackGenerator`, in this case capturing and recording system audio output to speakers or headphones https://github.com/guest271314/captureSystemAudio/blob/908210c941dfda4e237d09fb53a5f95bc7d25f49/native_messaging/capture_system_audio/background.js#L204.

Using `OfflineAudioContext`

```
        this.ac = new AudioContext({
          sampleRate: 44100,
          latencyHint: 0,
        });
        this.ac.suspend();
        this.msd = new MediaStreamAudioDestinationNode(this.ac, {
          channelCount: 2,
        });
        this.inputController = void 0;
        this.inputStream = new ReadableStream({
          start: (_) => {
            return (this.inputController = _);
          },
        });
        this.inputReader = this.inputStream.getReader();
        const { stream } = this.msd;
        this.stream = stream;
        const [track] = stream.getAudioTracks();
        this.track = track;
        this.osc = new OscillatorNode(this.ac, {
          frequency: 0,
        });
        this.processor = new MediaStreamTrackProcessor({
          track,
        });
        this.generator = new MediaStreamTrackGenerator({
          kind: 'audio',
        });
        const { writable: audioWritable } = this.generator;
        // ...
              this.audioReadable
              .pipeTo(
                new WritableStream({
                  abort(e) {
                    console.error(e.message);
                  },
                  write: async ({ timestamp }) => {
                    const int8 = new Int8Array(441 * 4);
                    const { value, done } = await this.inputReader.read();
                    if (!done) {
                      int8.set(new Int8Array(value));
                    } else {
                      console.log({done});
                      return this.audioWriter.closed;
                    }
                    const int16 = new Int16Array(int8.buffer);
                    // https://stackoverflow.com/a/35248852
                    const channels = [
                      new Float32Array(441),
                      new Float32Array(441),
                    ];
                    for (let i = 0, j = 0, n = 1; i < int16.length; i++) {
                      const int = int16[i];
                      // If the high bit is on, then it is a negative number, and actually counts backwards.
                      const float =
                        int >= 0x8000
                          ? -(0x10000 - int) / 0x8000
                          : int / 0x7fff;
                      // deinterleave
                      channels[(n = ++n % 2)][!n ? j++ : j - 1] = float;
                    }
                    // https://github.com/zhuker/lamejs/commit/e18447fefc4b581e33a89bd6a51a4fbf1b3e1660
                    const left = channels.shift();
                    const right = channels.shift();
                    let leftChannel, rightChannel;
                    if (this.mimeType.includes('mp3')) {
                      const sampleBlockSize = 441;
                      leftChannel = new Int32Array(left.length);
                      rightChannel = new Int32Array(right.length);
                      for (let i = 0; i < left.length; i++) {
                        leftChannel[i] =
                          left[i] < 0 ? left[i] * 32768 : left[i] * 32767;
                        rightChannel[i] =
                          right[i] < 0 ? right[i] * 32768 : right[i] * 32767;
                      }
                    }
                    const frame = new AudioData({
                      timestamp,
                      data: int16,
                      sampleRate: 44100,
                      format: 's16',
                      numberOfChannels: 2,
                      numberOfFrames: 441,
                    });
                    // ...
```

**Comment by AshleyScirra** [comment]
That code sample doesn't look relevant - it uses AudioContext, not OfflineAudioContext. I was talking about rendering audio offline, which is faster-than-realtime.

**Comment by guest271314** [comment]
What do you do with the rendered offline audio?

**Comment by guest271314** [comment]
Note, the same code can be used to encode faster than realtime; that is what the `OscillatorNode` is for, to get and set a `timestamp` on `AudioData` so playback does not occur too fast. You can encode media as fast as  `WritableStream` writes and a `ReadableStream` reads if you are not intending to stream the audio to speakers in parallel as you go.

**Comment by boourns** [comment]
@JohnWeisz From an API design perspective do you see value in @AshleyScirra's suggestion that `suspend()` return the partial audio buffer (or something similar)?  There seems to be agreement that it's important to get incremental audio buffers, it would be nice to have a solution that does not rely on doing it yourself with an `AudioWorkletNode`.

**Comment by JohnWeisz** [comment]
@boourns

Certainly, I'd say.

I may not use it myself, but in most cases, it's sure sounds more convenient than having to set up efficient cross-thread audio streaming like I explained.

**Comment by guest271314** [comment]
What is the source for the data used with `OfflineAudioContext`? An ordinary media file; dynamically created; a combination of both?

**Comment by hoch** [comment]
This work is actually one of WG's goals next year.

So far the consensus has been using ReadableStream along with OfflineAudioContext - as proposed in https://github.com/WebAudio/web-audio-api/issues/2445#issuecomment-844324592.

It needs a bit more clarification and the actual spec work. Some changes I have in mind:

```webidl
// From https://developer.mozilla.org/en-US/docs/Web/API/AudioData/format
enum AudioFormat {
  "u8",
  "s16",
  "s32",
  "f32",
  "u8-planar",
  "s16-planar",
  "s32-planar",
  "f32-planar"
}

dictionary OfflineAudioRenderingOptions {
  required AudioFormat format = "f32";
}

partial interface OfflineAudioContext {
  // Returns a stream that yields buffers of interleaved audio samples in Float32Array
  ReadableStream startRenderingStream(optional OfflineAudioRenderingOptions);  

  // If startRenderingStream() was called, this closes the stream and the context will be closed.
  Promise<void> stopRenderingStream();
};
```

The usage example:
```js
const context = new OfflineAudioContext(2, 44100, 44100);

// Add some nodes to build a graph...

const reader = context.startRenderingStream({format: "f32-planar"}).reader();

// An example of one iteration:
reader.read().then({done, value} => {
  if (done) {
    context.stopRenderingStream();
  }
  
  // Otherwise, do something with `value` and read() again.
});
```

Some notes:
1. It looks `value` in the reader's resolver can only be an object. Can we use a multi-dimension array?
2. The FooStream architecture is designed to serve "streaming" use cases. So there's a bit of disconnection between being offline (non-real time) and streaming (mostly for real time).
3. Because of 2, I still think about the old-fashion approach like `startRendering({length: someNumber})`. It's simple and less hassle.

**Comment by guest271314** [comment]
If I understand the requirement correctly `OfflineAudioContext.audioWorklet`, which processes faster than "real-time" can be used.

In `process()` post the `Float32Array` values to main thread using `port.postMessage()` or write to a `SharedArrayBuffer`.

**Comment by guest271314** [comment]
Re using `OfflineAudioContext.audioWorklet`, an example creating the stream from within the `AudioWorkletGlobalScope` context. The same approach applies when using `AudioBufferSourceNode` connected to `OfflineAudioContext`. In theory multiple `OfflineAudioContext.audioWorklet`s can be created and run in parallel. On Chromium 109 on *nix it takes ~1.5-~.1.5 seconds to create and stream 30 seconds of `Float32Array` data.

```
(async () => {
  const now = performance.now();
  const oac = new OfflineAudioContext({
    numberOfChannels: 2,
    length: 44100 * 30,
    sampleRate: 44100,
  });
  const worklet = `
class Stream extends AudioWorkletProcessor {
  constructor() {
    super();
  }
  process(inputs, outputs) {
    outputs[0][0].fill(Math.random() * 2 - 1);
    this.port.postMessage(outputs[0][0], [outputs[0][0].buffer]);
    return true;
  }
}

registerProcessor('stream', Stream);`;
  await oac.audioWorklet.addModule(
    URL.createObjectURL(
      new Blob([worklet], {
        type: 'text/javascript',
      })
    )
  );
  const floats = [];
  const aw = new AudioWorkletNode(oac, 'stream', {
    channelCount: 1,
    channelCountMode: 'explicit',
    numberOfInputs: 0,
    numberOfOutputs: 1,
    outputChannelCount: [1],
    processorOptions: {},
  });
  aw.port.onmessage = async (e) => {
    floats.push(...e.data);
  };
  aw.connect(oac.destination);
  await oac.startRendering();
  console.log((performance.now() - now) / 1000);
  console.log(floats);
  const ab = new AudioBuffer({
    length: 44100 * 30,
    sampleRate: 44100,
    numberOfChannels: 1,
  });
  ab.getChannelData(0).set(new Float32Array(floats));
  const ac = new AudioContext();
  const absn = new AudioBufferSourceNode(ac, {
    buffer: ab,
  });
  absn.connect(ac.destination);
  absn.start();
  absn.onend = async () => {
    await ac.close();
  };
})();
```

**Comment by hoch** [comment]
2/9 Teleconf:

> It looks value in the reader's resolver can only be an object. Can we use a multi-dimension array?

Perhaps this is the case where the interleaved stream format is better? (e.g. consumed by WebCodecs API)

- hongchan@ thinks `startRendering({length: someNumber})` also has an appeal.
- padenot@ wants to investigate the ReadableStream option further.

**Comment by hoch** [comment]
2023 TPAC Audio WG Discussion:

The discussion was brief, but the group generally agrees with the direction shown in https://github.com/WebAudio/web-audio-api/issues/2445#issuecomment-1277912417. However, the question remains:

- ReadableStream `value` is designed to be `Uint8Array`, so it might be not feasible to use it for interleaved or planar audio data in `Float32Array`.

@padenot any thoughts on this?

**Comment by guest271314** [comment]
This is how I do that

```
this.port.onMessage.addListener(async (value) => {
  try {
    await this.writer.ready;
    await this.writer.write(new Uint8Array(value));
  } catch (e) {
      console.log(e.message);
    } finally {
      return true;
    }
  });
```

```
this.stdout
  .pipeTo(
    new WritableStream({
        write: async (value, c) => {
          let i = 0;
          for (; i < value.buffer.byteLength; i++, this.readOffset++) {
            if (channelData.length === 441 * 4) {
              this.inputController.enqueue([...channelData]);
              channelData.length = 0;
            }
            channelData.push(value[i]);
          }
        },
```

```
this.audioReadable
  .pipeTo(
    new WritableStream({
        abort(e) {
          console.error(e.message);
        },
        write: async ({
            timestamp
          }) => {
            const int8 = new Int8Array(441 * 4);
            const {
              value,
              done
            } = await this.inputReader.read();
            if (!done) {
              int8.set(new Int8Array(value));
            } else {
              console.log({
                done
              });
              return this.audioWriter.closed;
            }
            const int16 = new Int16Array(int8.buffer);
            // https://stackoverflow.com/a/35248852
            const channels = [
              new Float32Array(441),
              new Float32Array(441),
            ];
            for (let i = 0, j = 0, n = 1; i < int16.length; i++) {
              const int = int16[i];
              // If the high bit is on, then it is a negative number, and actually counts backwards.
              const float =
                int >= 0x8000 ?
                -(0x10000 - int) / 0x8000 :
                int / 0x7fff;
              // Deinterleave
              channels[(n = ++n % 2)][!n ? j++ : j - 1] = float;
            }
```

I thing `MediaStreamTrackGenerator` and `MediaStreamTrackProcessor` achieves the requirement. Firefox has not implemented Media Capture Transform https://www.w3.org/TR/mediacapture-transform/. However, I don't think anything is stopping Web Audio API from implementing the same steps. 

The last time I checked Chrome `MediaStreamAudioDestinationNode` still does not produce silence per Media Capture and Streams, so we have to create an `OscillatorNode` for silence, then piggy-back on that to get `timestamp`s for WebCodecs `AudioData`. An extra step.

**Comment by padenot** [comment]
`MediaStreamTrackGenerator` and `MediaStreamTrackProcessor` aren't usable with `OfflineAudioContext`, please stay on topic. Generally, anything related to `MediaStream`s aren't usable with `OfflineAudioContext`, because `MediaStream` have to be associated are inherently real-time.

It's well understood how to convert an `UInt8Array` containing audio samples to floats, this is not the problem here.

**Comment by guest271314** [comment]
@padenot I said you can _model_ the solution here based on how `MediaStreamTrackProcessor` and `MediaSteamTrackGenerator` work. That is what you are trying to do here, write audio data t a `ReadableStream`.

**Comment by Vanilagy** [comment]
Hey, is there any update on this? I am combining OfflineAudioContext with Web Codecs, and the fact that it's impossible to incrementally retrieve the audio chunks as they're rendering is a major limiting factor - either I have a huge AudioBuffer and large latency as the encoder can only start once the renderer is finished, or I need to do some gymnastics and split up my rendering into multiple, contiguous OfflineAudioContexts, which makes the code complex and also breaks down as soon as I start using convolutions or delays. Is there a current "best practice" for doing incremental/progressive rendering via OfflineAudioContext?

I do think the ReadableStream approach presented by @hoch is fine, although my favorite API suggestion is probably just `startRendering({ length: numberOfFrames })`, which returns an AudioBuffer of that length and advances the context's currentTime by that amount of frames.

**Comment by bartadaniel** [comment]
@Vanilagy For this very reason, I ended up moving most of my audio processing to a custom WASM-based solution. I’m still wondering if there’s an easier way to work around this nowadays.

This may not be the best place to bring it up, but issues like this current one, [this](https://github.com/WebAudio/web-audio-api/issues/2423), [this](https://issues.chromium.org/issues/40942807), and [this](https://github.com/WebAudio/web-audio-api/issues/2442) made me move most of my code away from the Web Audio API. It’s clear that a lot of thought and hard work went into designing this API and it’s really just a few issues away from being the ultimate audio API. But unfortunately, some of these have remained open for quite some time now.

**Comment by clevertree** [comment]
Disappointing but expected from planned obsolescence. The global digital audio workstation (DAW) market was valued at USD 2.95 billion in 2023 and is projected to reach USD 6.34 billion by 2032. They were not going to let WebAudioAPI this be the reason that money isn't got. Sorry, has to be said.

**Comment by Vanilagy** [comment]
@bartadaniel Using a custom, WASM-based solution is cool. Thinking about also going that route. Sucks that I have to consider it in the first place, since the browser has such a vast and natively optimized audio API, but a few issues block it from being truly useful.

**Comment by Vanilagy** [comment]
@JohnWeisz I'll try this out, thanks. Should solve the streaming issue, with the only remaining issue being the large AudioBuffer that is allocated no matter what (and that I have to predetermine the duration of the output buffer). That's still 1.4 GB for a 1-hour render.

**Comment by Vanilagy** [comment]
Using the idea by @JohnWeisz, I implemented progressive rendering using OfflineAudioContext, and it actually works very well.

I tried both AudioWorklet and the deprecated ScriptProcessorNode, and sorry to say, but ScriptProcessorNode is totally superior here. Not only is the code much simpler, but it was overall *more* performant while being free of any race conditions. Makes sense: The ScriptProcessorNode simply passes its input audio buffer to the encoder, and writes nothing to the output, so it has essentially zero overhead and there's nothing wrong with running it on the main thread.

It also required careful usage of suspend/resume on OfflineAudioContext. My final rendering loop follows this pattern now:
```ts
let lastSuspendPromise = audioContext.suspend(0); // Queue suspense before starting to render audio
const renderingDone = audioContext.startRendering(); // This starts rendering, but halts immediately

for (let currentFrame = 0; currentFrame < totalFrames; currentFrame++) {
    const currentTime = currentFrame / frameRate;

    if (currentFrame > 0 && currentFrame % frameRate === 0) { // Every second
        await lastSuspendPromise; // In case the last suspense time isn't reached yet, wait

        lastSuspendPromise = audioContext.suspend(currentTime);
        void audioContext.resume(); // Render audio up until currentTime
    }

    // Update the scene, perform any rendering, queue new audio nodes, ...
    // ...
}

await lastSuspendPromise;
void audioContext.resume();
await renderingDone;
```

This pattern is clean. It allows me to concurrently render video and audio, throttling if the video renders faster than the audio (which can happen in my case).

Perfect. Performant, progressive audio rendering. Only issue now is the unnecessary allocation happening for the output buffer returned by `startRendering`.

---

One big problem though: This doesn't work on Firefox. Firefox doesn't have support for suspend/resume on OfflineAudioContext, and without these, I'm pretty sure the above logic cannot be modeled. What a shame! I guess it's always possible to fall back to simply using `startRendering`'s output buffer, but it does add annoying if-else logic to the code.

**Comment by Vanilagy** [comment]
@JohnWeisz 

> I think this is a misunderstanding, you don't really have to call suspend for progressive rendering to work via ScriptProcessorNode/AudioWorkletNode at all. Just keep emitting your sample chunks as they come.

Yes, for pure rendering, this is true. The added value of suspend/resume is the ability to build your audio graph incrementally *while* rendering is taking place. This is especially relevant in my case where, to determine the audio graph, I first need to iterate over my entire timeline to "simulate" my scene. Doing that and doing the rendering only at the end is equivalent to using the output from `startRendering` (although encoding is faster if I do it chunked). Therefore, it's very useful to incrementally render out the audio before the full graph is even built yet. I also notice this has a substantial performance improvement beyond just chunking: audio rendering time appears to scale with the number of nodes in the graph, and rendering it in a chunked way means I can only have, let's say, 1000 nodes around at a given time rather than the 20000 I would need for the full render. Much better for perf and memory.

Your suggested fix by blocking in an AudioWorklet is actually genius, though! I mean, almost feels hacky to me, but it should be equivalent in behavior to suspend/resume if I implement it right. I'll dig into that. Might be the way to polyfill suspend/resume for Firefox!

**Comment by mattbirman** [comment]
> 2023 TPAC Audio WG Discussion:
> 
> The discussion was brief, but the group generally agrees with the direction shown in [#2445 (comment)](https://github.com/WebAudio/web-audio-api/issues/2445#issuecomment-1277912417). However, the question remains:
> 
> * ReadableStream `value` is designed to be `Uint8Array`, so it might be not feasible to use it for interleaved or planar audio data in `Float32Array`.
> 
> [@padenot](https://github.com/padenot) any thoughts on this?

Has there been any progress on this since the 2023 TPAC Audio WG discussion?

**Comment by mattbirman** [comment]
The use case we're targeting is video processing. We used to use WebAudio for realtime audio processing but couldn't use the OfflineAudioContext because of the limitations described in this issue. Ideally, we'd like to use WebAudio for realtime and offline processing.

I'm happy to contribute to the discussion and spec work. What would be the next steps?

**Comment by hoch** [comment]
This would be my general approach:

1. Summarize the discussion on this topic.
2. Enumerate several designs.
3. Draft an explainer based on the research (1+2).
4. Review the explainer with the Audio WG and TAG. (iterate as needed)
5. Draft a PR for the spec. (iterate as needed)
6. Merge the PR and start the implementation.

Once you have an explainer ready, we can definitely invite you to our bi-weekly conference and have a discussion.

Thanks for your interest and feel free to ask me any questions!

**Comment by gabrielsanbrito** [comment]
@hoch @padenot @mattbirman Here is a summary from the [TPAC 2025 meeting's minutes](https://docs.google.com/document/d/15bLL2UujSW4OdSa0Kv66wziXBTq34jWOntNDbs2pPek/edit?tab=t.dpp97ymwtdhx):
- Feature detection is a must.
- Support for a `chunk size` is a good idea.
- Need to reach consensus on what format to use. Ideally support multiple formats. 

To get started on this, I think that we can:
- Create a new method `startRenderingStream` to make the feature detectable.
- This method will accept a dictionary of parameters, which will also me it easier to add new parameters in the future without changing the API surface.
- One of the dictionary parameters can be `chunkSize`. **What unit should be used for it? Time (milliseconds) or number of frames? Something else?**
- To start, should we start supporting only interleaved 8-bit data? If I understood correctly, it should be simpler to encode the output stream using Web Codecs. Later on we can introduce support for new formats and augment the dictionary to also have a `format` parameter.

For this first iteration the API could look like this. What do you think?

```js
dictionary OfflineAudioRenderingOptions {
  // The number of frames to render each iteration
  Number chunkSize = 128;
}

partial interface OfflineAudioContext {
    // Immediately stops the rendering, to implement a "cancel" button when rendering 
    // if startRenderingStream was called, this closes the stream
    // If startRendering was called, this rejects the promise
    Promise<void> close();
    // Returns a stream that yields buffers of interleaved audio samples in Float32Array or whatever format is specified
    Promise<ReadableStream> startRenderingStream(optional OfflineAudioRenderingOptions);
};
```

We have updated the proposed approach in our [explainer](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/OfflineAudioContext/explainer.md#proposed-approach---add-startrenderingstream-function).

**Comment by JohnWeisz** [comment]
In the meantime, a few thoughts on that excellent summary in https://github.com/WebAudio/web-audio-api/issues/2445#issuecomment-3663099259

 > To start, should we start supporting only interleaved 8-bit data? If I understood correctly, it should be simpler to encode the output stream using Web Codecs. Later on we can introduce support for new formats and augment the dictionary to also have a format parameter.

To me, it sounds like the most natural format would be planar float32, i.e. a `ReadableStream<Float32Array[]>` -- simply given how everything else in the Web Audio API uses it, including the `AudioBuffer` returned by `startRendering`.

We have everything on the web platform today to trivially make that into an interleaved byte stream through a TransformStream.

 > One of the dictionary parameters can be chunkSize. What unit should be used for it? Time (milliseconds) or number of frames? Something else?

I'm thinking number of frames, given you typically reason about a stream in chunk byte sizes. Definitely shouldn't be a time in milliseconds though, given everything else in the Web Audio API uses seconds (`AudioBuffer.duration`, `BaseAudioContext.currentTime`, `AudioWorkletGlobalScope.currentTime`).

-----

Something I didn't see mentioned but is going to be a potential problem: **what about backpressure?** It's going to be important if your OfflineAudioContext graph is simple and your encoding pipeline is expensive, or if you are e.g. streaming to a remote destination over a slow network, etc. as long as it's slower than your offline rendering speed. Without backpressure, memory usage may build up and lead to an OOM crash.

My homegrew polyfill handles this by simply suspending the OfflineAudioContext until the destination is ready to receive more chunks, but we probably don't want that to happen unexpectedly.

**Comment by Vanilagy** [comment]
Love how this new API is shaping up! A question: What about the constructor? Currently, we're required to construct the `OfflineAudioContext` with a fixed length (in number of samples). If rendering can be streamed in the future, I can see a need for audio contexts unbounded in length as well. The regular online AudioContext is unbounded as well, so I see no reason why this one can't be either.

```ts
const context = new OfflineAudioContext({
  numberOfChannels: 2,
  sampleRate: 48000,
  // 'length' omitted
});
```

In this case, the `ReadableStream` could just generate chunks ad infinitum until `close()` is called on the context.

**Comment by karlt** [comment]
There has been discussion around another new means to get audio data from the AudioDestinationNode and some of these involve new methods to control suspend and resume of the context:

The resume/suspend effect of `startRendering(length)` is perhaps clear.

`startRenderingStream()` would also resume and suspend the context, because it would need to respond to back-pressure: the ReadableStream internal queue is not exposed to the client, so the client cannot respond to back-pressure.

ReadableStream on its own is somewhat vague about the when exactly the underlying source produces each chunk, but Web Audio could specify that the  OfflineAudioContext underlying source is a pull source with no read-ahead.
This is particularly relevant when the rendering graph is being changed between chunks.

If an OfflineAudioContext constructed with no `length` has a different means to control suspend and resume, then it would be less confusing if predetermined-length OfflineAudioContext `suspend()` and `resume()` methods were simply not available on the arbitrary-length OfflineAudioContext.

A key question then is: does the ReadableStream API provide sufficient benefit over other methods (AudioWorklet, AnalyserNode, [buffers from resume(length) or statechange events](https://github.com/WebAudio/web-audio-api/issues/2662#issuecomment-3748096147)) that it is worthwhile replacing the `suspend(suspendTime)`/`resume()` API, for arbitrary-length OfflineAudioContext?

**Comment by hoch** [comment]
* The WG discussed the `OfflineAudioContext` streaming capabilities, concluding that using the `length` parameter to change the object's behavior was not a viable approach.
* Two design proposals will be developed for comparison: one retrofitting the existing `OfflineAudioContext` and another creating a new "streaming" or "on-demand" audio context class.
* @hoch and @gabrielsanbrito will create a document to present these ideas, including sketches for a new context design and drafts for the retrofitted original context.

**Comment by gabrielsanbrito** [comment]
This is the [document](https://docs.google.com/document/d/1g8RQEyrrt-b_5Em3-7iMGyhoG4ZZmz51Q-9QNdfHpLc/edit?tab=t.0#heading=h.vyy2rjn8zqhz) mentioned by @hoch, which has been already shared with the WG (the proposal below has not been added there yet as I am writing this).

After the last WG meeting there has been some preference towards a callback-based model that returns chunks through a promise<AudioBuffer>.  The current proposal is:

```webidl
dictionary OfflineAudioContextOptions {
    // not required, defaults to render quantum size. 
    unsigned long length;  

    // Members below are unchanged.
    unsigned long numberOfChannels = 1;
    required float sampleRate;
    (AudioContextRenderSizeCategory or unsigned long) renderSizeHint = "default";
};

partial interface OfflineAudioContext {
    Promise<void> close();
    Promise<AudioBuffer> startRendering(optional unsigned long chunkSize);
}
```

In this proposal:
- `OfflineAudioContextOptions.length` is optional and defaults to the render quantum.
- If unspecified, `chunkSize` defaults to `OfflineAudioContextOptions.length`.
- All offline audio rendering is "progressive" by definition:
  - The current one-shot rendering scenario becomes a special case where `OfflineAudioContextOptions.length` is specified and `chunkSize` is not specified (defaults to `OfflineAudioContextOptions.length`).
  - Unknown duration rendering is supported by not providing an `OfflineAudioContextOptions.length` value. Progressive rendering can be done by calling `startRendering` multiple times. If `chunkSize` is not provided, it defaults to `OfflineAudioContextOptions.length`, which should be the render quantum size in this case.
  - A new `close()` method was added. This should behave similarly to `AudioContext.close()`, and, once the returned promise resolves, the `OfflineAudioContext` should move to the `closed` state and become unusable.

 The rationale behind this proposal is:
- ReadableStream adoption has been dropped due to the added complexity of dealing with it, both from the implementors and users perspective. For more context, read the [Decoupling WebCodecs from Streams](https://docs.google.com/document/d/10S-p3Ob5snRMjBqpBf5oWn6eYij1vos7cujHoOCCCAw/edit?tab=t.0#heading=h.xiy0ernna2ml) document.
- By using callbacks, application can control audio rendering more precisely and it makes usage simpler.
- We don't need a `startRenderingStream()` method because we can:
  - Allow `startRendering()` to be called multiple times
  - Modify `startRendering()` to `startRendering(optional chunkSize)`
  - Fallback to rendering a single chunk with length given by `OfflineAudioContextOptions.length` if `chunkSize` is not specified (current behavior).
- `OfflineAudioContextOptions.length` is optional to support scenarios where we don't know in advance how much audio we need to render
- An explicit `close()` method is required to release resources of "hanging" `OfflineAudioContexts`. This is important for when an `OfflineAudioContext` is rendering a track of indefinite length - i.e. when `OfflineAudioContextOptions.length` is not specified.
- BYOB support has been dropped from the proposal because it would introduce unnecessary complexity.
  - If the rendering is complex enough to require an audio context, then an allocation for each new audio chunk will not be significant overhead.
- In the interest of focusing on the iterative audio rendering problem, I think we can drop the format discussion for now. 

Please let us know what you think. Thanks!

Edit 1: added missing `unsigned long` type for `chunkSize`.
Edit 2: added the explicit `close()` method.

**Comment by gabrielsanbrito** [comment]
@hoch @padenot @chrisguttandin I just updated my previous comment to also include an explicit `close()` method based on the latest WG meeting feedback. With this, I think we can start drafting the required spec changes. What do you think?

**Comment by hoch** [comment]
Thanks for your patience, @gabrielsanbrito. Overall it looks good, but I want to make a few things clear.

- Concurrent Calls to startRendering(): If they call it concurrently (without waiting the promise to be resolved), should the calls be queued, or should it throw an error? 
- Currently, OfflineAudioContext has a readonly `length` attribute.  If `length` is omitted in the options and defaults to a small render quantum size (e.g., 128), what does the length attribute on the context instance return? Also does it return dynamic values as rendering progresses?
- In the indefinite rendering scenario (where length is omitted), how does the system or the user know that the rendering is "done"?
- The existing spec fires an `oncomplete` event when the entire rendering is finished. In the progressive case, does this event still fire?

After answering these questions, I think we can move onto PR drafting and the review. What do you think @padenot?

**Comment by karlt** [comment]
Please speak up if you see things differently, but these were my expectations:

> * Concurrent Calls to startRendering(): If they call it concurrently (without waiting the promise to be resolved), should the calls be queued, or should it throw an error?

I'm not perceiving a strong demand for rendering ahead, so I suggest starting with throwing an error, because that keeps things simpler.
Concurrent calls can be added later if deemed appropriate.

> * Currently, OfflineAudioContext has a readonly `length` attribute.  If `length` is omitted in the options and defaults to a small render quantum size (e.g., 128), what does the length attribute on the context instance return? Also does it return dynamic values as rendering progresses?

If the `length` option is omitted, `length` returns the render quantum size (e.g. 128).
This is not dynamic.  It indicates the default for the `chunkSize` parameter.
`currentTime` is available as a measure of total rendering performed.

> * In the indefinite rendering scenario (where length is omitted), how does the system or the user know that the rendering is "done"?

Rendering is done when the Promise from the most recent startRendering() has settled and the associated oncomplete has fired.

Omitting the `length` merely means to use the render quantum size as the default for `chunkSize`,  and provides feature detection for progressive rendering.  It doesn't change other behavior: indefinite rendering is still available even if the `length` is provided.

> * The existing spec fires an `oncomplete` event when the entire rendering is finished. In the progressive case, does this event still fire?

I think the behavior most consistent with the current spec would be to fire oncomplete after each and every startRendering().

**Comment by gabrielsanbrito** [comment]
@hoch thanks for raising these clarifying questions! @karlt, I agree partially with you. After these questions I am starting to think that making `length` required while allowing it to be set to `Infinity` might make more sense. This is how I believe it should work based on the group discussion so far:

>> Concurrent Calls to startRendering(): If they call it concurrently (without waiting the promise to be resolved), should the calls be queued, or should it throw an error?

> I'm not perceiving a strong demand for rendering ahead, so I suggest starting with throwing an error, because that keeps things simpler.
Concurrent calls can be added later if deemed appropriate.

We would like to have render-ahead from the get-go and I believe that they should be queued.

>> Currently, OfflineAudioContext has a readonly length attribute. If length is omitted in the options and defaults to a small render quantum size (e.g., 128), what does the length attribute on the context instance return? Also does it return dynamic values as rendering progresses?

> If the length option is omitted, length returns the render quantum size (e.g. 128).
This is not dynamic. It indicates the default for the chunkSize parameter.
currentTime is available as a measure of total rendering performed.

@karlt I know that this also works, but, "semantically" I find it hard to grasp the meaning of "length" if it means "default chunkSize". In my understanding, "length" is the amount of data the context will process before being `"closed"`. I believe that it would make sense to return `Infinity` in this case. This is similar to [HTMLMediaElement.duration](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/duration) and possibly communicates intent clearer?

I propose making `OfflineAudioContextOptions.length` required again and allowing it to be set to `Infinity`. This could help avoid users creating infinite duration contexts by "accident".

- `OfflineAudioContextOptions.length` is allowed to be set to be `Infinity`.
- If unspecified, chunkSize defaults to OfflineAudioContextOptions.length. If OfflineAudioContextOptions.length is `Infinity`, make chunkSize equal to the render quantum.
- All offline audio rendering is "progressive" by definition:
  - The current one-shot rendering scenario becomes a special case where OfflineAudioContextOptions.length is not `Infinity` and chunkSize is not specified (defaults to OfflineAudioContextOptions.length).
  - Unknown duration rendering is supported by making OfflineAudioContextOptions.length equal to `Infinity`. Progressive rendering can be done by calling startRendering multiple times. If chunkSize is not provided, it defaults to:
    - The render quantum size if OfflineAudioContextOptions.length is `+inf`.
    - OfflineAudioContextOptions.length otherwise

What do you think?

>> In the indefinite rendering scenario (where length is omitted), how does the system or the user know that the rendering is "done"?

> Rendering is done when the Promise from the most recent startRendering() has settled and the associated oncomplete has fired.

I think @hoch is interested in knowing when a signal will be sent to the platform to release resources. I think that the caller should decide when the rendering is finished by calling `close()`.

>> The existing spec fires an oncomplete event when the entire rendering is finished. In the progressive case, does this event still fire?
> I think the behavior most consistent with the current spec would be to fire oncomplete after each and every startRendering().

It makes sense to fire oncomplete after each and every startRendering(). But, for the [OfflineAudioCompletionEvent](https://webaudio.github.io/web-audio-api/#OfflineAudioCompletionEvent), I see in the spec that: "This is an [Event](https://dom.spec.whatwg.org/#event) object which is dispatched to [OfflineAudioContext](https://webaudio.github.io/web-audio-api/#OfflineAudioContext) for legacy reasons".

I think it should still fire in the defined-length case to preserve backwards compatibility. 

In the **indefinite**/"infinite" duration case I think we shouldn't since there isn't a need for it and it seems that we are just keeping this event around for legacy reasons.

**Comment by karlt** [comment]
> , "semantically" I find it hard to grasp the meaning of "length" if it means "default chunkSize"

Semantically, the name "length" doesn't distinguish between a sum of the lengths of multiple rendering chunks or the length of one rendering operation.
There is flexibility available to choose what the parameter describes.

> I believe that it would make sense to return Infinity in this case. This is similar to HTMLMediaElement.duration and possibly communicates intent clearer?

> I propose making OfflineAudioContextOptions.length required again and allowing it to be set to Infinity. This could help avoid users creating infinite duration contexts by "accident".

Is there a concern with having all OfflineAudioContexts capable of continuing to render if startRendering() calls are repeated?
Is there a benefit to having a mechanism to constrain at context construction the total frames that a particular OfflineAudioContext could render through one or more startRendering() calls?
Is that benefit sufficiently valuable, that it justifies an extra explicit required parameter to indicate when no such maximum is intended?

Is this a concern about memory resources remaining in the AudioNodes?
If startRendering() has recently completed, then it will have just produced an AudioBuffer, which will presumably consume memory resources at least until the next GC.  Is the resource use of AudioNodes expected to be so much greater that it needs to be released immediately after rendering instead of at the next GC?

* I could imagine a DelayNode with a large buffer, but presumably this would not usually be large relative to the output of rendering.

* Gecko makes no attempt to release resources from AudioNodes of OfflineAudioContexts before GC.  Do other browsers release earlier?

* Constraining the total frames wouldn't mean earlier release of resources in a context that has rendered less than a maximum total length.  close() would still be required for that.  Would the close() method be sufficient to make a constraint on total frames unnecessary?

If both a maximum total length and a chunkSize could be specified, then there would be two potentially contradictory indications of how much to render.  The spec could state which takes priority when the chunkSize would indicate more than the maximum total length, but it would remain a source of potential confusion for clients.

**Comment by gabrielsanbrito** [comment]
Yes. But, I am not thinking just about `AudioBuffers`. I am thinking also about the broader meaning of [system resources](https://webaudio.github.io/web-audio-api/#releasing) outlined in the Web Audio spec. My concern was mostly about these "resources" living on unnecessarily.

When the length is predefined, the OfflineAudioContext "knows" that it won't render anything beyond the provided length and can execute something similar to the `AudioContext.close()` [steps](https://webaudio.github.io/web-audio-api/#dom-audiocontext-close). Whereas, if the length is not provided, the "close" steps will run only if `close()` is explicitly called or when the context is destroyed.

By enabling the indefinite-length rendering scenario when `length` is not provided, we make it easier to create these indefinite-length OfflineAudioContexts. If this behavior is available only via `length = Infinity`, users need to be intentional about it. 

> If both a maximum total length and a chunkSize could be specified, then there would be two potentially contradictory indications of how much to render. The spec could state which takes priority when the chunkSize would indicate more than the maximum total length, but it would remain a source of potential confusion for clients.

I don't see this as being a big problem. Let `n` be the total number of frames rendered so far at any point during an OfflineAudioContext rendering session. The next rendered chunk size `cs` will be given by:
`cs = (n + cs < length) ? cs : length - n`

**Comment by hoch** [comment]
I understand both arguments, but based on my experience with AudioContext's resource usage and the problems I've seen, I think having a `close()` method without a defined `length` is my personal choice. Memory/usage management is in the domain of the browser implementation, and the spec discussion should place more value on usability and developer experience.

@padenot WDYT?

**Comment by gabrielsanbrito** [comment]
@hoch @padenot closing this now that the feature has been added to the spec.
